geomstats/geometry/sasaki_metric.py

```diff
@@ -0,0 +1,277 @@
+"""Class for the Sasaki metric.
+
+A class implementing the Sasaki metric: The natural metric on the tangent bundle TM
+of a Riemannian manifold M.
+
+Lead authors: E. Nava-Yazdani, F. Ambellan, M. Hanik and C. von Tycowicz.
+"""
+import os
+
+from joblib import Parallel, delayed
+
+import geomstats.backend as gs
+from geomstats.geometry.riemannian_metric import RiemannianMetric
+from geomstats.tests import tf_backend
+
+N_STEPS = 3
+
+
+def _gradient_descent(x_ini, i_pt, e_pt, grad, exp, lrate=0.1, max_iter=100, tol=1e-6):
+    """Apply a gradient descent until max_iter or a given tolerance is reached."""
+    x = x_ini
+    for _ in range(max_iter):
+        grad_x = grad(x, i_pt, e_pt)
+        grad_norm = gs.linalg.norm(grad_x)
+        if grad_norm < tol:
+            break
+        grad_x = -lrate * grad_x
+        x = exp(grad_x, x)
+    return x
+
+
+class SasakiMetric(RiemannianMetric):
+    """Implements of the Sasaki metric on the tangent bundle TM of a Riem. manifold M.
+
+    The Sasaki metric is characterized by the following three properties:
+     * the canonical projection of TM becomes a Riemannian submersion,
+     * parallel vector fields along curves are orthogonal to their fibres, and
+     * its restriction to any tangent space is Euclidean.
+
+    Geodesic computations are realized via a discrete formulation of the geodesic
+    equation on TM that involve geodesics, parallel translation, and the curvature
+    tensor on the base manifold M (see [1] for details).
+    However, as the implemented energy in the discrete-geodesics-optimization as well as
+    the approximations of its gradient slightly differ from those proposed in [1], we
+    also refer to [2] for additional details.
+
+    Parameters
+    ----------
+    metric : RiemannianMetric
+        Metric of the base manifold of the tangent bundle.
+
+    References
+    ----------
+    .. [1] Muralidharan, P., & Fletcher, P. T. (2012, June).
+    Sasaki metrics for analysis of longitudinal data on manifolds.
+    In 2012 IEEE conference on computer vision and pattern recognition (pp. 1027-1034).
+    https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4270017/
+    [2] Nava-Yazdani, E., Hanik, M., Ambellan, F., & von Tycowicz, C. (2022, June).
+    On Gradient Formulas in an Algorithm for the Logarithm of the Sasaki Metric
+    Technical Report Zuse-Institut Berlin (no. ZIB-22-12) (urn:nbn:de:0297-zib-87174).
+    https://nbn-resolving.org/urn/resolver.pl?urn:nbn:de:0297-zib-87174
+    """
+
+    def __init__(self, metric: RiemannianMetric):
+        self.metric = metric
+        shape = (2, gs.prod(gs.array(metric.shape)))
+
+        super(SasakiMetric, self).__init__(
+            2 * metric.dim, shape=shape, default_point_type="matrix"
```

Comment:
`default_point_type` should be `self.metric.default_point_type`. 

I know this will raise problems in Frechet mean. I'lll open a PR to solve that issues there (still today). I think it is the same problem we've been discussing with shape.

Reply:
@lpereira95, well I think `default_point_type=self.metric.default_point_type` is not a good idea as it potentially would introduce wrong configurations. If we e.g. take an input metric with `self.metric.default_point_type="vector"`the SasakiMetric will have 'shape = (2, gs.prod(gs.array(metric.shape)))' that, as I understand, suggests that `default_point_type="matrix"`. In fact, in the discussion about the shape (above) I wrote (to argue why 'shape = (2, gs.prod(gs.array(metric.shape)))' is a good idea):

> ... vectorizing the shape shape of the base manifold has the advantage that the iteration of taking the sasaki_metrix (i.e. looking at the tangent bundle of the tangent bundle) does not introduce new dimensions to the shape.

This essentially says that we always get the same `default_point_type` and that is "matrix".

Long story short: I would prefer keeping the current code here, as long as SasakiMetric vectorizes its input.